smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -587,6 +608,36 @@ private String getNamedMembersInputParam(
         }
     }
 
+    /**
+     * Given context and a source of data, generate an input value provider for the
+     * map.
+     *
+     * @param context The generation context.
+     * @param bindingType How this value is bound to the operation input.
+     * @param dataSource The in-code location of the data to provide an input of
+     *                   ({@code input.foo}, {@code entry}, etc.)
+     * @param target The shape of the value being provided.
+     * @return Returns a value or expression of the input collection.
+     */
+    private String getMapInputParam(
+            GenerationContext context,
+            Location bindingType,
+            String dataSource,
+            MapShape target
+    ) {
+        Model model = context.getModel();
+        MemberShape mapMember = target.getValue();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+
+        String valueString = getInputValue(context, bindingType, "value", mapMember,
+                model.expectShape(mapMember.getTarget()));
+        return "Object.entries(" + dataSource + " || {}).reduce("
+            + "(acc: any, [key, value]: [string, " + symbolProvider.toSymbol(mapMember) + "]) => ({"
+            +   "...acc,"
+            +   "[key]: " + valueString + ","
+            + "}), {})";
```

Comment:
I get why you did it this way - this should work for any sort of map you throw at it. At the same time, the query params trait can only ever be applied to `map<string, string>` or `map<string, list<string>>` which just so happens to be [the precise type signature of your query map](https://github.com/aws/aws-sdk-js-v3/blob/main/packages/types/src/http.ts#L60). It seems like it would be much simpler to just directly expand it into the query map if it's defined. That would turn this:

```typescript
  const query: any = {
    ...(input.queryParamsMapOfStrings !== undefined &&
      Object.entries(input.queryParamsMapOfStrings || {}).reduce(
        (acc: any, [key, value]: [string, string]) => ({ ...acc, [key]: value }),
        {}
      ))
  }
```

into this:

```typescript
  const query: any = {
    ...(input.queryParamsMapOfStrings !== undefined ? input.queryParamsMapOfStrings : {})
  }
```

or

```typescript
const query: any = {};
if (input.queryParamsMapOfStrings !== undefined) {
  query = { ...input.queryParamsMapOfStrings, ...query }
}
```

It's worth noting that the trait is structurally exclusive, meaning there will only ever be one member bearing it on a given operation.

Reply:
Fixed in [`45f371c` (#311)](https://github.com/awslabs/smithy-typescript/pull/311/commits/45f371c48552e524f7e5b935ed9e3d811dc5153f)

Codegen the following output code:
```js
  const query: any = {
    ...(input.queryParamsMapOfStrings !== undefined && input.queryParamsMapOfStrings)
  }
```

Verified the new api-gateway client in https://github.com/aws/aws-sdk-js-v3/pull/2249